linkml/generators/pydanticgen/array.py

```diff
@@ -0,0 +1,396 @@
+    the supported pydantic versions (currently still 1 and 2).
+
+    Notes:
+
+        When checking for array specification, recall that there is a semantic difference between
+        ``None`` and ``False`` , particularly for :attr:`.ArrayExpression.max_number_dimensions` -
+        check for absence of specification with ``is None`` rather than checking for truthiness/falsiness
+        (unless that's what you intend to do ofc ;)
+    """
+
+    REPR: ClassVar[ArrayRepresentation]
+
+    def __init__(self, array: ArrayExpression | None, dtype: str | Element, pydantic_ver: str = PYDANTIC_VERSION):
+        self.array = array
+        self.dtype = dtype
+        self.pydantic_ver = pydantic_ver
+
+    def make(self) -> SlotResult:
+        """Create the string form of the array representation"""
+        if not self.array.dimensions and not self.has_anonymous_dimensions:
+            # any-shaped array
+            return self.any_shape(self.array)
+        elif not self.array.dimensions and self.has_anonymous_dimensions:
+            return self.anonymous_shape(self.array)
+        elif self.array.dimensions and not self.has_anonymous_dimensions:
+            return self.labeled_shape(self.array)
+        else:
+            return self.mixed_shape(self.array)
+
+    @property
+    def has_anonymous_dimensions(self) -> bool:
+        """Whether the :class:`.ArrayExpression` has some shape specification aside from ``dimensions``"""
+        return any([getattr(self.array, arr_field, None) is not None for arr_field in _ANONYMOUS_ARRAY_FIELDS])
+
+    @classmethod
+    def get_generator(cls, repr: ArrayRepresentation) -> Type["ArrayRangeGenerator"]:
+        """Get the generator class for a given array representation"""
+        for subclass in cls.__subclasses__():
+            if subclass.REPR == repr:
+                return subclass
+        raise ValueError(f"Generator for array representation {repr} not found!")
+
+    @abstractmethod
+    def any_shape(self, array: Optional[ArrayRepresentation] = None) -> SlotResult:
+        """Any shaped array!"""
+        pass
+
+    @abstractmethod
+    def anonymous_shape(self, array: ArrayExpression) -> SlotResult:
+        """Array shape specified numerically, without axis parameterization"""
+        pass
+
+    @abstractmethod
+    def labeled_shape(self, array: ArrayExpression) -> SlotResult:
+        """Array shape specified with ``dimensions`` without additional anonymous axes"""
+        pass
+
+    @abstractmethod
+    def mixed_shape(self, array: ArrayExpression) -> SlotResult:
+        """Array shape with both ``dimensions`` and a ``max_number_dimensions`` for anonymous axes"""
+        pass
+
+
+class ListOfListsArray(ArrayRangeGenerator):
+    """
+    Represent arrays as lists of lists!
+
+    TODO: Move all validation of values (eg. anywhere we raise a ValueError) to the ArrayExpression
+    dataclass and out of the generator class
+    """
+
+    REPR = ArrayRepresentation.LIST
```

Comment:
I'm curious - what is this for? Currently the classes map 1:1 to representations. Is the point of this variable to allow multiple classes to return the same array representation?

Reply:
It was a way of having a firm notion of what representations exist for the purposes of testing, and also used by the generator to know which representations to emit, but it is overly restrictive in that it doesnt allow for additional representations to hook in as plugins. I think it should probably be replaced by introspecting `__subclasses__`

Reply:
Also 1:1 class:representation is what we want. If someone wants to patch a representation we should make hooks, but didnt want to try and tackle that in this one PR, but good eye, agreed its not ideal

Reply:
Ah, I understand and also see the helper function `get_generator` on line 209 above now. I think this is fine for now.